Assignments/Program.cs

```diff
@@ -1 +1,110 @@
+      for (int i = 0; i < 3; i++) {
+         // Here part is FIRST 3-digits for (i = 0), SECOND 3-digits for (i = 1), and LAST 3-digits for (i = 2).
+         // Eg: num = 123456 ==> input = 000123456.  Then, part = 000 (for i = 0); 123 (for i = 1); 456 (for i = 2).
+         int part = int.Parse (input.Substring (i * 3, 3));
+         // For any 3-digit number, the place value of first digit is always Hundreds.  So the state s always starts with Hundreds.
+         State s = State.Hundreds;
+         // Below loop is used to take out digits ONE BY ONE from part and converted into words.
+         for (int n = 3; n > 0; n--) {
+            // nthDigit is the 3rd digit of part for (n = 3), and so on.
+            // onesDigit is the always the 3rd digit of part.
+            int nthDigit = part / (int)Math.Pow (10, n - 1) % 10, onesDigit = part % 10;
+            // The below switch cases is implemented based on the STATE MACHINE DIAGRAM attached in "Data".
+            switch (s, nthDigit) {
+               case (State.Hundreds, >= 0 and <= 9): {
+                     s = State.Tens;
+                     if (nthDigit != 0) {
+                        output += ones[nthDigit] + " Hundred ";
+                        if (part / 10 % 10 > 0 || onesDigit > 0) output += "and ";
+                     }
+                     break;
+                  }
+               case (State.Tens, 1): {
+                     s = State.End;
+                     output += ones[10 + onesDigit] + " ";
+                     break;
+                  }
+               case (State.Tens, 0 or (>= 2 and <= 9)): {
+                     s = State.Ones;
+                     if (nthDigit != 0) output += tens[nthDigit] + ((i == 2 && onesDigit > 0) ? "-" : " ");
+                     break;
+                  }
+               case (State.Ones, >= 0 and <= 9): {
+                     s = State.End;
+                     if (nthDigit != 0) output += ones[nthDigit] + " ";
+                     break;
+                  }
+               default: { s = State.End; break; }
+            }
+         }
+         if (part > 0 && i != 2) {
+            var (cond, key) = (i == 0) ? (num / 1000 % 1000 > 0 || num % 1000 > 0, "Million") : (num % 1000 > 0, "Thousand");
+            output += key + (cond ? ", " : " ");
+         }
+      }
+      return output;
+   }
+   enum State { Hundreds, Tens, Ones, End };
+
+   /// <summary>Converts input number to Roman Numerals</summary>
+   public static string ToRomans (int num) {
+      List<string> output = new ();
+      int[] value = { 1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1 };
+      string[] roman = { "M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I" };
+      for (int i = 0; i < value.Length; i++) {
+         int x = value[i];
+         output.Add (string.Concat (Enumerable.Repeat (roman[i], num / x)));

```

Comment:
If I want to convert the number 1 to Roman this list has 12 empty strings and finally the Roman letter for 1. Can you think of reducing the iterations?